src/limbo/experimental/model/spgp.hpp

```diff
@@ -0,0 +1,628 @@
+//| with a limited warranty  and the software's author,  the holder of the
+//| economic rights,  and the successive licensors  have only  limited
+//| liability.
+//|
+//| In this respect, the user's attention is drawn to the risks associated
+//| with loading,  using,  modifying and/or developing or reproducing the
+//| software by the user in light of its specific status of free software,
+//| that may mean  that it is complicated to manipulate,  and  that  also
+//| therefore means  that it is reserved for developers  and  experienced
+//| professionals having in-depth computer knowledge. Users are therefore
+//| encouraged to load and test the software's suitability as regards their
+//| requirements in conditions enabling the security of their systems and/or
+//| data to be ensured and,  more generally, to use and operate it in the
+//| same conditions as regards security.
+//|
+//| The fact that you are presently reading this means that you have had
+//| knowledge of the CeCILL-C license and that you accept its terms.
+//|
+
+//| TODO: Add credits to the original implementation
+#ifndef LIMBO_MODEL_SPGP_HPP
+#define LIMBO_MODEL_SPGP_HPP
+
+#include <cassert>
+#include <iostream>
+#include <limits>
+#include <vector>
+
+#include <Eigen/Cholesky>
+#include <Eigen/Core>
+#include <Eigen/LU>
+
+#include <limbo/tools.hpp>
+#include <limbo/opt.hpp>
+
+namespace limbo {
+    namespace model {
+        /// @ingroup model
+        /// A classic Gaussian process.
+        /// It is parametrized by:
+        /// - a mean function
+        /// - [optionnal] an optimizer for the hyper-parameters
+        template <typename Params, typename KernelFunction, typename MeanFunction, class HyperParamsOptimizer = opt::NLOptGrad<Params, nlopt::LD_LBFGS>>
+        class SPGP {
+        public:
+
+            struct HyperParams
+            {
+                Eigen::MatrixXd b;
+                double c;
+                double sig;
+                Eigen::MatrixXd xb;
+
+                HyperParams(const Eigen::VectorXd& input, const size_t& m, const size_t& dim_in) {
+                    Eigen::VectorXd w = input;
+                    b =  w.segment(m*dim_in, dim_in).transpose().array().exp();
+                    c = std::exp(w[(m+1)*dim_in]);
+                    sig = std::exp(w[(m+1)*dim_in+1]);
+                    xb = Eigen::Map<Eigen::MatrixXd>(w.segment(0, m).data(), m, dim_in);
+                }
+            };
+
+            /// useful because the model might be created before knowing anything about the process
+            SPGP() : _dim_in(-1), _dim_out(-1) {}
+
+            /// useful because the model might be created before having samples
+            SPGP(int dim_in, int dim_out)
+                : _dim_in(dim_in), _dim_out(dim_out) {}
+
+            SPGP(const Eigen::MatrixXd& samples, const Eigen::MatrixXd& observations, 
+                size_t m, double jitter = 0.000001, double dim_out = 1)
```

Comment:
> Shouldn't jitter be in the Params struct?

It should. This is still **WIP**, so please find as many typos/things to change as possible. Thanks!